File: template-mirror/atlas-api/src/features/identity/infrastructure/json-identity-store.ts
import { readFileSync } from "node:fs";
import { randomBytes } from "node:crypto";
import type { BenchUser } from "../domain/user";
import { toEvaluationContext } from "../domain/user";
import type { IdentityStore } from "../application/ports/identity-store";

type SeedFile = { users?: BenchUser[] };

export function createJsonIdentityStore(input: {
  enabled: boolean;
  seedPath: string;
}): IdentityStore {
  const users = input.enabled ? loadUsers(input.seedPath) : [];
  const byId = new Map(users.map((u) => [u.id, u]));
  const sessions = new Map<string, string>();

  return {
    enabled: input.enabled,
    listUsers(limit) {
      return users.slice(0, Math.max(1, limit));
    },
    listByOrg(org, limit) {
      return users
        .filter((u) => u.org === org)
        .slice(0, Math.max(1, limit));
    },
    login(userId) {
      const user = byId.get(userId);
      if (!user) return null;
      const token = randomBytes(16).toString("hex");
      sessions.set(token, user.id);
      return { token, user, evaluationContext: toEvaluationContext(user) };
    },
    session(token) {
      if (!token) return null;
      const userId = sessions.get(token);
      if (!userId) return null;
      const user = byId.get(userId);
      if (!user) return null;
      return { user, evaluationContext: toEvaluationContext(user) };
    },
  };
}

function loadUsers(seedPath: string): BenchUser[] {
  try {
    const raw = JSON.parse(readFileSync(seedPath, "utf8")) as SeedFile;
    return Array.isArray(raw.users) ? raw.users : [];
  } catch {
    return [];
  }
}

File: apps/cove-orders/src/features/identity/infrastructure/json_identity_store.py
import json
import secrets
from dataclasses import asdict
from pathlib import Path
from typing import Optional

from src.features.identity.domain.user import BenchUser, to_evaluation_context


def create_json_identity_store(*, enabled: bool, seed_path: str):
    users = _load_users(seed_path) if enabled else []
    by_id = {u.id: u for u in users}
    sessions: dict[str, str] = {}
    store_enabled = enabled

    class Store:
        enabled = store_enabled

        def list_users(self, limit: int) -> list[dict]:
            return [asdict(u) for u in users[: max(1, limit)]]

        def login(self, user_id: str):
            user = by_id.get(user_id)
            if not user:
                return None
            token = secrets.token_hex(16)
            sessions[token] = user.id
            return {
                "token": token,
                "user": asdict(user),
                "evaluationContext": to_evaluation_context(user),
            }

        def session(self, token: Optional[str]):
            if not token:
                return None
            user_id = sessions.get(token)
            if not user_id:
                return None
            user = by_id.get(user_id)
            if not user:
                return None
            return {
                "user": asdict(user),
                "evaluationContext": to_evaluation_context(user),
            }

    return Store()


def _load_users(seed_path: str) -> list[BenchUser]:
    try:
        raw = json.loads(Path(seed_path).read_text(encoding="utf-8"))
        rows = raw.get("users") if isinstance(raw, dict) else None
        if not isinstance(rows, list):
            return []
        out: list[BenchUser] = []
        for row in rows:
            if not isinstance(row, dict):
                continue
            out.append(
                BenchUser(
                    id=str(row.get("id", "")),
                    email=str(row.get("email", "")),
                    name=str(row.get("name", "")),
                    org=str(row.get("org", "")),
                    plan=row.get("plan", "free"),
                    country=str(row.get("country", "")),
                )
            )
        return out
    except (OSError, json.JSONDecodeError):
        return []
